### DISCUSSION TITLE
Traffic

### DISCUSSION THREAD

**Discussion by ItsX12** [organicmaps/organicmaps#568] [Ideas]
I would like a feature to be able to see traffic on different roads.

**Comment by GLLM** [comment]
I'd love this, but I'm not sure there are currently reliable traffic source than are free and can be used for FOSS projects.
Is there any ?

**Reply by ItsX12** [reply]
I know there is a project called OpenTraffic (https://github.com/opentraffic) I don't know if this could be of use.

**Comment by dikkechill** [comment]
I really like to see this feature as well, as it might bring organicmaps towards the feature level of google maps for routing.
And hence make it a real alternative.

The graphhopper project has a nice curated list of public traffic sources:
* https://github.com/graphhopper/open-traffic-collection/blob/master/README.md
* https://www.graphhopper.com/blog/2015/04/08/visualize-and-handle-traffic-information-with-graphhopper-in-real-time-for-cologne-germany-koln/

The issue I see is that standards differ per region/country and importing and maintaining this would be cumbersome.
Potentially something like opentraffic.io (https://github.com/opentraffic) could help standardize things, but not sure how active this project still is.

Another alternative is to integrate one or more paid services, where organicmaps users could sign up themselves.
* https://www.mapbox.com/traffic-data/ (seems to be enterprise only)
* https://distancematrix.ai/product

Some other ideas:
* Estimate default values for roads / times, see a discussion on https://github.com/Ifsttar/NoiseModelling/discussions/303
* Provide an API within OrganicMaps for people to develop their own traffic plugins (allowing the community to develop their own regional/country imports based on demand, making maintenance easier)
* Team up with other projects who have similar requirements, e.g. OSMAnd (https://github.com/osmandapp/OsmAnd/issues/6878)
* Check how other OSM based apps do it, e.g. MagicEarth (https://www.magicearth.com/feature-availablity/)

**Reply by citizenunique** [reply]
Real time traffic information is the feature I miss the most in organicmaps and the reason why I'm still using maps.me, to be honest. 
I also found opentraffic.io (https://github.com/opentraffic) and think it could the solution?

And I agree also that a paid service could be an alternative way, I at least would definitively pay as an in-app purchase in organicmaps for this feature.

Thanks to the developers!

**Reply by biodranik** [reply]
Does real-time traffic still work in MapsMe? Opentraffic project looks dead, unfortunately.

We are looking for the easiest and simplest way to solve this issue. If users could pay for the traffic info directly to the provider, it would be great.

**Reply by citizenunique** [reply]
Real-time traffic is still available in Maps.me 10.2.3.0, which I'm still using for this reason. It works great.
Since this is not available anymore in the current Maps.me version, this could be a killer feature if available in Organic Maps.

**Reply by TheAdventurer64** [reply]
How did MAPS.ME do real time traffic?

**Reply by biodranik** [reply]
They made a business deal with some proprietary traffic provider.

**Reply by citizenunique** [reply]
This is how it looked like in the old version:

![6F8995DE-74CC-4617-A3F9-35FAD8E757A1](https://user-images.githubusercontent.com/33265052/154099350-290a4cda-31b9-4311-bf63-8b556af38b1d.jpeg)

**Reply by borsj** [reply]
Maybe it is possible to create an user input to add an online overlay map to the map output. So the user can create a google traffic overlay with the right link.

**Comment by Nosnahc** [comment]
Magic Earth has a traffic service that looks open source
https://www.magicearth.com/crowd-sourced-traffic/

**Reply by biodranik** [reply]
Crowd-sourced means that they collect it from their users. It doesn't mean that it's open-source (the source code is available to everyone).

**Reply by Nosnahc** [reply]
My bad

**Comment by b3n63r** [comment]
A basic traffic feature would be very useful, like not routing into scheduled closed roads due to construction. I have closed road in my town and it is marked highway=construction in OSM, but organic maps doesn't care.

This is the only thing keeping me from using organic maps more often for car navigation.

**Reply by Misalf-git** [reply]
Seems to work here.
I just tryed to get routed through a road tagged with both `highway=construction` and `construction=residential` but Organic Maps correctly sent me on a detour, avoiding the road under construction.

Maybe you want to open an issue with coords and other related info.

**Reply by b3n63r** [reply]
It may be the used OSM version is per 21th of January (if I recall it correctly) and the change to construction is per 15th of February. Is this possible?

**Reply by biodranik** [reply]
If the edit in osm was done AFTER we generated maps, then of course routing won't know about it.

**Comment by Benger85** [comment]
What about the good old TMC? The data is produced by public organisations in many countries and it seems that it is now available via internet, so there should be a way to use it in OrganicMaps for traffic: https://tisa.org/technologies/tpeg/

**Reply by biodranik** [reply]
Are there any APIs to get this information for all supported countries?

**Reply by Benger85** [reply]
I am not an expert. I couldn't find any API but this https://tisa.org/tpeg-standards-and-recommended-use/

Looks like most of the documents in the webshop are free. I have no idea what you might be interested in, but hopefully it will help.

**Reply by dikkechill** [reply]
TPEG seems mainly used for premium (paid) services, https://en.wikipedia.org/wiki/TPEG.
E.g. TomTom provides it's HD traffic data in this format for carmakers and manufacturers (https://www.tomtom.com/products/real-time-traffic/). TomTom also offers it in OpenLR (http://www.openlr.org/).
Additionally they offer data in Datex II and protobuf (https://developer.tomtom.com/intermediate-traffic-service/documentation/product-information/introduction)

**Comment by dikkechill** [comment]
One alternative mentioned in https://github.com/organicmaps/organicmaps/issues/1160 is TraFF (https://gitlab.com/traffxml)
Benefit is that the project is open source and uses public (free) data. It provides a number of implementations (TMC and Datex II to TraFF, XML server for data feed subscriptions, example implementations). Impressive for a one person project and it seems easily extendable.
Drawback is that there is no infrastructure running for it yet and many countries are not covered yet. At the moment it's also a one person project.

**Comment by PinguDEV-original** [comment]
Over the weekend i built my own project, https://codeberg.org/pinguin/open-traffic-data .
Currently one-person, currently in a custom GeoJSON format. Will develop this a bit more tho.
Data is being updated once per minute and can be found here: https://traffic.derpingu.in/open-traffic-data.geojson

**Reply by jamessucla** [reply]
Just pointing out this resource for Americans since I don't see it mentioned anywhere else: Sigalert.com (owned by iHeartMedia) has freely available live traffic speed data for every major city, which could probably be converted to GeoJSON.

Using the Southern California region as an example:
Speeds: https://www.sigalert.com/Data/SoCal/0~j/SoCalData.json
Labels and Coordinates(?): https://cdn-static.sigalert.com/240/Zip/RegionInfo/SoCalStatic.json

**Comment by PinguDEV-original** [comment]
I found TomTom, they have a free plan (2.5k requests per day) and offer live traffic data for many places. What I'd say: TomTom should be integrated into OMaps but without a set API key so a user can register by themself. But in the future we could support like a "pro" plan (2$ a month or so) where user don't have to get the API key themselves but can pay for it. I'd say make the latter option customizable, so anyone can host a service to register (and to pay money to).
Please comment your ideas, because maybe it's just bullshit what i'm thinking..

**Comment by hansaplasst** [comment]
In the Netherlands the [Ministry of Infrastructure and Water Management ](https://rijkswaterstaatdata.nl/data-zoeken/#hd2d7d2cb-a48e-4a32-8ca2-45a278d8073e) publishes Road traffic data via it's [Nationaal Dataportaal Wegverkeer](https://www.ndw.nu/). You can use this data without a license. You don't even need a login. "Realtime" data is available for download via a compressed [XML file ](https://opendata.ndw.nu/gebeurtenisinfo.xml.gz)

The dataset covers the following categories:
- Dynamic traffic signs and rules
- Bridge Closures and Conditions of Access
- Lane Closures and Access Conditions
- Other temporary traffic measures or plans
- Road closures and access conditions
- Others, other things, or others
- Real-time traffic data
- Speed and speed
- Transport volume of traffic
- Road conditions and events
- Accidents and incidents
- Poor road conditions
- Weather conditions on the road
- Information about road works
- Long-term roadworks
- Short road works
- Static road signs and rules
- Other traffic rules

Would be very nice if someone could implement this.

**Comment by zachvlat** [comment]
there is still this google traffic that can be implemented with this link: http://mts0.google.com/vt/lyrs=h,traffic&x={1}&y={2}&z={0}&style=3](http://mts0.google.com/vt/lyrs=h,traffic&x=%7B1%7D&y=%7B2%7D&z=%7B0%7D&style=3

**Comment by NAVY-GreenArrow** [comment]
Hello @biodranik is there any detailed information on how to build MWM maps? I tried the docs and what a mess, the source code is full of .so, but I can't see how to create it. I'm also trying to use a local .osm.pbf on my PC since I can't afford much network consumption. Also, many tests require a complete uninstallation of OM, and that forces me to download maps again, which I can't. If a team member could help me with this, that would be great.

**Comment by FunctionalHacker** [comment]
Looks like someone created a live traffic feature with the previously mention TomTom API in osmandapp/OsmAnd#25356. It can do routing around traffic jams. Perhaps Organic Maps can take some inspiration from there.
